Integrate Plotly into OED

CorinaSterbet opened the pull request:
This pull request replaces the old graphics package Chart.js with the new Plotly.js. 

The code for all graph types (line, bar, compare) has been remodeled/rewritten to correspond to the requirements of the new graphics package. Some notes:
- the line graph now features a new option to zoom in and out on specific time intervals chosen by the user;
- this pull request solves issue #392 which caused, for the bar graph, the values and meter names to be mismatched at the graph display;
- I have looked into issue #312 -- however, I could not detect the errors mentioned there. When I checked the week value and compared it against the sum of the day values, they came up the same (which is what we want).

Commit:
Merge pull request #1 from OpenEnergyDashboard/development

Update fork

Commit:
Merge remote-tracking branch 'origin/fast-pt-compression' into development

Commit:
Changes to packages after "fast pt" merge

Commit:
Add the plotlyjs-ts types

Commit:
Update package.json with plotly packages

Commit:
Display the plotly line graph with fast pt data

Commit:
The Plotly line graph works with fast pt

*changed the line chart container, so that the compressed readings are passed to the x and y values of the new chart
*the dashboard container and component have been changed to feature the "Redraw" and "Restore" buttons
*added several comments

Commit:
Minor edit

Commit:
Place the buttons on the line chart page only

And get rid of the buttons from the bar chart and compare pages.

Commit:
Convert the bar graph to Plotly

*fix the dashboard component warning
*rename the linechar variable
*convert the old bar graph code to Plotly
*fix the bar bug: the old code pushes the  bar readings to the data variable in the order received, and then takes the timestamps (labels) and orders them chronologically. This is an issue since the actual readings do not get ordered with the time intervals associated with them.

Commit:
Clean up the code from the debug statements

Commit:
The compare graph works with plotly

At this point, the 'Week' and '4 Week' graphs are displayed without any problems. The 'Day' graph is not loading and I suspect that that will be solved when the pull request "compare with shift" is intergated into my branch.

Commit:
Merge remote-tracking branch 'upstream/development' into plotly-and-upstream-dev

Commit:
Add a type to plotlychart to fix errors

Commit:
Finish the code for compare graphs

*the code for compare is now in plotly and works with the latest version of the dev
*fixed a bug related to the display of how many % more or less have been consumed in the current period when compared to the last period (compare graphs)

Commit:
Fix  the labels for compare

huss commented on src/client/app/containers/LineChartContainer.ts:
```diff
@@ -54,73 +78,75 @@ function mapStateToProps(state: State) {
-			duration: 0
+	// Customize the layout of the plot
+	const layout: any = {
+		autozise: true,
```
```suggestion
		autosize: true,
```
Wrong name/typo. autosize as per <https://plot.ly/javascript/reference/#layout-autosize>

huss commented on src/client/app/types/react-plotlyjs-ts/index.d.ts:
```diff
@@ -0,0 +1,30 @@
+/// <reference types="react" />
```
```suggestion
/// <reference types="react" />
// TODO <https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html> gives limited 
// circumstances where reference should be used. We need to verify this is a good use.
```
<https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html> gives limited circumstances where reference should be used. Wanted to check that this is the best way to do this.

huss commented on src/client/app/containers/LineChartContainer.ts:
```diff
@@ -54,73 +78,75 @@ function mapStateToProps(state: State) {
+				const xData: string[] = [];
+				const yData: number[] = [];
+				const hoverText: string[] = [];
+				_.values(readingsData.readings).forEach(reading => {
```
```suggestion
				_.values(readingsData.readings).forEach(reading => {
// TODO The meter code explicitly creates the array but this does not. This seems cleaner (a little harder to 
// see data on debug). Whatever is chosen, it should probably be the same way.
```

huss commented on src/client/app/containers/LineChartContainer.ts:
```diff
@@ -31,19 +24,50 @@ function mapStateToProps(state: State) {
+					line: {
+						shape: 'spline',
+						width: 3
+						// smoothing: 1.3
```
```suggestion
```
Seems left over item so remove at this point.

huss commented on src/client/app/containers/LineChartContainer.ts:
```diff
@@ -54,73 +78,75 @@ function mapStateToProps(state: State) {
+	// Customize the layout of the plot
+	const layout: any = {
+		autozise: true,
+		// width: 700,
```
Should these 3 lines be removed?

huss commented on src/client/app/containers/LineChartContainer.ts:
```diff
@@ -54,73 +78,75 @@ function mapStateToProps(state: State) {
+		// width: 700,
+		// height: 700,
+		// title: 'First Test',
+		showlegend: true,
```
```suggestion
		showlegend: true,
// Just for the record, showlegend is listed as true by default in the Plotly documentation but it turns out 
// it does not show the first one unless you set it this way so this is needed.
```

huss commented on src/client/app/containers/LineChartContainer.ts:
```diff
@@ -54,73 +78,75 @@ function mapStateToProps(state: State) {
-			}
+		yaxis: {
+			title: 'kW',
+			showgrid: true,
```
```suggestion

```
It is okay to set showgrip: true but that is the default. Testing shows grids are present without this line (or in xaxis). Given we don't include all the defaults, it is suggest to remove this.

huss commented on src/client/app/containers/LineChartContainer.ts:
```diff
@@ -54,73 +78,75 @@ function mapStateToProps(state: State) {
-		data,
-		options,
-		redraw: true
+	// Assign all the paramaters required to create the Plotly object (data, layout, config) to the variable props, returned by mapStateToProps
```
```suggestion
	// Assign all the parameters required to create the Plotly object (data, layout, config) to the variable props, returned by mapStateToProps
```
paramaters -> paramaters

huss commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -11,60 +11,132 @@ import MultiCompareChartContainer from '../containers/MultiCompareChartContainer
+import { FormattedMessage } from 'react-intl';
+import { Dispatch, Thunk, ActionType } from '../types/redux/actions';
+
+import * as Plotly from 'plotly.js';
```
```suggestion
import * as Plotly from 'plotly.js';
// TODO lowercase plotly for the import in index.d.ts but uppercase here.
```

huss commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -11,60 +11,132 @@ import MultiCompareChartContainer from '../containers/MultiCompareChartContainer
 
-	const optionsClassName = props.optionsVisibility ? 'col-2 d-none d-lg-block' : 'd-none';
-	const chartClassName = props.optionsVisibility ? 'col-12 col-lg-10' : 'col-12';
+	public render(){
```
```suggestion
	public render() {
```
space before {

huss commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -11,60 +11,132 @@ import MultiCompareChartContainer from '../containers/MultiCompareChartContainer
+
+	private handleTimeIntervalChange(mode: string) {
+		if (mode == "all"){
+			this.props.changeTimeInterval(TimeInterval.unbounded());			
```
```suggestion
			this.props.changeTimeInterval(TimeInterval.unbounded());
```
3 tabs on end of line

huss commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -11,60 +11,132 @@ import MultiCompareChartContainer from '../containers/MultiCompareChartContainer
+	private handleTimeIntervalChange(mode: string) {
+		if (mode == "all"){
+			this.props.changeTimeInterval(TimeInterval.unbounded());			
+		}else{
```
```suggestion
		} else {
```
spacing on } and {

huss commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -11,60 +11,132 @@ import MultiCompareChartContainer from '../containers/MultiCompareChartContainer
+			let sliderBox: any = document.querySelector(".rangeslider-slidebox");
+			let root: any = document.getElementById("root");	
+
+			if (sliderContainer && sliderBox && root){
```
```suggestion
			if (sliderContainer && sliderBox && root) {
```
space before {

huss commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -11,60 +11,132 @@ import MultiCompareChartContainer from '../containers/MultiCompareChartContainer
+
+				let timeInterval = new TimeInterval(moment(newMinXTimestamp), moment(newMaxXTimestamp));
+				this.props.changeTimeInterval(timeInterval);
+			}			
```
```suggestion
			}
```
3 tabs at end of line

huss commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -11,60 +11,132 @@ import MultiCompareChartContainer from '../containers/MultiCompareChartContainer
+		if (mode == "all"){
+			this.props.changeTimeInterval(TimeInterval.unbounded());			
+		}else{
+			let sliderContainer: any = document.querySelector(".rangeslider-bg");
```
This series of calculations seems to work fine and makes some sense. What I have been unable to find/figure out is where the rangeslider-bg & rangeslider-slidebox are set/documented. I did find <https://plot.ly/javascript/reference/#layout-xaxis-rangeslider> on Plotly but that does not match this. I suspect I'm missing something basic but wanted to be sure this is absolutely correct.

huss replied:
I've decided to let this go and close this out.

huss commented on src/client/app/components/DashboardComponent.tsx:
```diff
@@ -11,60 +11,132 @@ import MultiCompareChartContainer from '../containers/MultiCompareChartContainer
+		);		
+	}
+
+	private handleTimeIntervalChange(mode: string) {
```
```suggestion
	private handleTimeIntervalChange(mode: string) {
// TODO Should mode be an enum since it has clearly defined values. Since the if/else uses it below 
// it could use a boolean but that implies there will only be two possible values forever 
// which seems likely to be the case.
```

huss commented on src/client/app/containers/BarChartContainer.ts:
```diff
@@ -23,18 +21,26 @@ function mapStateToProps(state: State) {
 			const readingsData = byMeterID[timeInterval.toString()][barDuration.toISOString()];
 			if (readingsData !== undefined && readingsData.readings !== undefined && !readingsData.isFetching) {
```
This file does && readingsData.readings !== undefined whereas LineChartContainers.ts has a separate if with a thrown Error. Would be good to be consistent unless there is a reason not to be. Same for groups below.

huss replied:
Code modified to do this.

huss commented on src/client/app/containers/BarChartContainer.ts:
```diff
@@ -23,18 +21,26 @@ function mapStateToProps(state: State) {
 			if (readingsData !== undefined && readingsData.readings !== undefined && !readingsData.isFetching) {
 				const label = state.meters.byMeterID[meterID].name;
-				const color = getGraphColor(label);
+				const xData: string[] = [];
```
A number of comments were added to LineChartContainer.ts and it might be nice to have the mirror ones here.

huss replied:
Addressed in commit 62e075f

huss commented on src/client/app/containers/LineChartContainer.ts:
```diff
@@ -31,19 +24,50 @@ function mapStateToProps(state: State) {
+				const hoverText: string[] = [];
+				const readings = _.values(readingsData.readings);
+				readings.forEach(reading => {
+					const timeReading = moment(reading.startTimestamp);
```
The code here differs slighly from BarCharContainer.ts. They do the same basic work so it would be nice if they did it in a similar way.

huss replied:
Addressed in commit 62e075f

huss commented on src/client/app/containers/BarChartContainer.ts:
```diff
@@ -3,18 +3,16 @@
-	const datasets: ChartDataSets[] = [];
+	const datasets: any[] = [];
 
 	const labelsSet = new Set();
```
```suggestion
```
This appears left over from the old previous version so remove.

huss commented on src/client/app/containers/BarChartContainer.ts:
```diff
@@ -45,87 +51,77 @@ function mapStateToProps(state: State) {
+		barmode: 'group',
+		bargap: 0.2,
+		bargroupgap: 0.1,
+		autozise: true,
```
autosize not autozise.

huss replied:
```suggestion
		autosize: true,
```
Wrong name/typo.

huss commented on src/client/app/containers/BarChartContainer.ts:
```diff
@@ -45,87 +51,77 @@ function mapStateToProps(state: State) {
-		data,
-		options,
-		redraw: true
+	// Assign all the paramaters required to create the Plotly object (data, layout, config) to the variable props, returned by mapStateToProps
```
paramaters -> parameters

huss replied:
```suggestion
	// Assign all the parameters required to create the Plotly object (data, layout, config) to the variable props returned by mapStateToProps
```

huss commented on src/client/app/containers/BarChartContainer.ts:
```diff
@@ -45,87 +51,77 @@ function mapStateToProps(state: State) {
+			},
+			tickangle: -45,
+			autotick: true,
+			nticks: 10,
```
<https://plot.ly/javascript/reference/#scattergl-marker-colorbar-tickmode> does not seem to make it clear if tickmode is auto (at least not from my very modest checking). nticks only has impact in this case. Should we set tickmode?

huss replied:
Upon rereading, it states for the other two cases what is the default so it seems it is likely okay in this case. Since working, I am going to close this.

huss commented on src/client/app/containers/CompareChartContainer.ts:
```diff
@@ -41,107 +28,76 @@ function mapStateToProps(state: State, ownProps: CompareChartContainerProps): Li
+			color: colorize(entity.change)
 		},
+		hovermode:'closest',
+		autozise: true,
```
```suggestion
		autosize: true,
```
autosize

huss commented on src/client/app/containers/CompareChartContainer.ts:
```diff
@@ -41,107 +28,76 @@ function mapStateToProps(state: State, ownProps: CompareChartContainerProps): Li
+			t: 20,
+			b: 120,
+			l: 60,
+			r: 20,
```
In the Plotly reference at <https://plot.ly/javascript/reference/#layout-margin>, I cannot find the r attribute for margin. Want to verify this is correct usage.

nicolebryant replied:
I noted a similar issue but found that the attribute is seen in the search function of plotly's documentation although not found in the documentation itself. I have put in an issue on Plotly's Documentation GitHub regarding this.

huss replied:
Let's assume it is okay to use, esp. since not causing any issues.

huss commented on src/client/app/containers/LineChartContainer.ts:
```diff
@@ -54,73 +78,75 @@ function mapStateToProps(state: State) {
+		autozise: true,
+		// width: 700,
+		// height: 700,
+		// title: 'First Test',
```
```suggestion
```
Seems left over from development so remove.

huss committed:
Update src/client/app/containers/LineChartContainer.ts

Remove old comments.

huss committed:
Apply suggestions from code review

Small change fixups.

huss commented on src/client/app/containers/LineChartContainer.ts:
```diff
@@ -31,19 +24,49 @@ function mapStateToProps(state: State) {
 			}
 		}
 	}
 
```
```suggestion

// TODO The meters and groups code is very similar and maybe it should be refactored out to create a function to do 
// both. This would mean future changes would automatically happen to both.
```

huss committed:
Apply suggestions from code review

Add TODO for these items or do simple change.

huss committed:
Make check in Bar/Line ChartContainer the same

huss committed:
address some PR comments

Adds commnets to Bar from Line ChartContainer
Error if undefined reads as in Line (did before for meters)
In line, creates variable with readings as in bar so
debug easier and consistent.

huss committed:
Merge branch 'development' of origin

Fixes merge conflicts with development so PR should be okay.
Ignores chartjs changes to fix # digits
Used most recent version of either package in package.json
Regenerated package-lock.json so correct for all these files.

huss commented:
The code in this PR does what it was designed to do. However, there are some other features we want before we deploy the new Plotly graphics system for general use. See issues #456, #458, #459 & #460. This is just to document this fact.

nicolebryant committed:
Remove all Chart.js/chartjs/chart-js references in code

nicolebryant committed:
Update README.md

huss committed:
update package-lock.json

The accomplishes a copuple of things:
- The changes Nicole made caused some packages to use SHA1 not SHA512.
This is a known issue and this puts many at SHA 512.
- We had many outdates and vulnerable packages which have been updated.
Note there are still 20 packages that need to be looked at.

The only known issue that some developer may see is that the upgrades
need a newer TypeScript (believed to be at least 3.5)
or they will get errors about Omit
but OED still runs.
It is also noted that there are many continuing lint errors that will
be addressed (not new to these changes).

nicolebryant committed:
Remove lint errors from DashboardComponent

nicolebryant committed:
Fix lint erros in BarChartContainer.ts

nicolebryant committed:
Fix lint errors in CompareChartContainer.ts

nicolebryant committed:
Fix lint errors in LineChartContainer.ts

nicolebryant committed:
Attempt to update index.d.ts

huss committed:
Merge branch 'fix-lint-errors' of https://github.com/nicolebryant/OED into pr/CorinaSterbet/436

This is to help address the linting errors.

huss committed:
Resolve linting issues

Remove leftover "
Tabs for spaces, declare class members public
Remove comment at top since item removed.

huss commented:
This is to document that I verified all graphic data (line with day and hourly, all bar & all compare) against the raw meter data. They were in agreement but some edge issues were found that will be new issues and outside the work of this PR. This should be a strong test that fast-pt and the new graphics in Plotly are doing what they are supposed to do.

huss commented:
The remaining TS lint error:
TypeError: ts.unescapeIdentifier is not a function
seems likely to be related to the versions of tslint/tsutil we are using per <https://github.com/palantir/tslint/issues/4266>. We want to migrate to ESLint (#463) so hope it goes away then. This is not causing an error in the Travis build so not resolving now.

huss committed:
Fix missing header

nicolebryant commented:
This PR looks good to merge based on the discussions of code reviews and testing with @huss as well as the updates made to the code. There are still some minor changes that need to be made later that aren't breaking or causing the Travis build to fail, all listed in the comments on this PR.

huss reviewed (approved):
This PR looks good to merge based on the discussions of code reviews and testing with @huss as well as the updates made to the code. There are still some minor changes that need to be made later that aren't breaking or causing the Travis build to fail, all listed in the comments on this PR
